### PR TITLE
Mirror verifies received STH a bit more thoroughly.

### PR DESCRIPTION
**Pull request by AlCutter**
Addresses #761

### CODE REVIEW

File: cpp/server/ct-mirror.cc [hunk 1]
```diff
@@ -134,6 +135,7 @@ using std::shared_ptr;
 using std::string;
 using std::thread;
 using std::unique_ptr;
+using util::HexString;
```
Comment (pphaneuf) [inline cpp/server/ct-mirror.cc]: Not used? And I'm guessing that the `#include "util/util.h"` is for this, so unneeded?

Comment (AlCutter) [inline cpp/server/ct-mirror.cc]: Yeah, used in an earlier iteration of the PR.
removed.

File: cpp/server/ct-mirror.cc [hunk 2]
```diff
@@ -218,11 +220,13 @@ static const bool follow_dummy =
...
     Database<LoggedCertificate>* db,
     ClusterStateController<LoggedCertificate>* cluster_state_controller,
-    mutex* queue_mutex, map<int64_t, ct::SignedTreeHead>* queue, Task* task) {
+    mutex* queue_mutex, map<int64_t, ct::SignedTreeHead>* queue,
+    LogLookup<LoggedCertificate>* log_lookup, Task* task) {
   CHECK_NOTNULL(db);
   CHECK_NOTNULL(cluster_state_controller);
   CHECK_NOTNULL(queue_mutex);
   CHECK_NOTNULL(queue);
+  CHECK_NOTNULL(log_lookup);
```
Comment (pphaneuf) [inline cpp/server/ct-mirror.cc]: And what do you do with this `log_lookup`?

File: cpp/fetcher/remote_peer.h [hunk 3]
```diff
@@ -2,11 +2,15 @@
 #define CERT_TRANS_FETCHER_REMOTE_PEER_H_
 
 #include "fetcher/peer.h"
+#include "log/log_lookup.h"
 #include "log/log_verifier.h"
 #include "util/task.h"
 
+
```
Comment (pphaneuf) [inline cpp/fetcher/remote_peer.h]: Whoa there! :smiley:

File: cpp/fetcher/remote_peer.cc [hunk 4]
```diff
@@ -101,10 +148,12 @@ void RemotePeer::Impl::DoneGetSTH(
 
 RemotePeer::RemotePeer(
     unique_ptr<AsyncLogClient>&& client, unique_ptr<LogVerifier>&& verifier,
+    LogLookup<LoggedCertificate>* log_lookup,
     const std::function<void(const ct::SignedTreeHead&)>& on_new_sth,
     util::Task* task)
     : Peer(move(client)),
-      impl_(new Impl(move(verifier), client_.get(), on_new_sth, task)) {
+      impl_(new Impl(move(verifier), client_.get(), log_lookup, on_new_sth,
```
Comment (pphaneuf) [inline cpp/fetcher/remote_peer.cc]: Add a `CHECK_NOTNULL`?

File: cpp/server/ct-mirror.cc (hunk 2 above)
Comment (AlCutter) [inline cpp/server/ct-mirror.cc]: same

File: cpp/fetcher/remote_peer.cc (hunk 4 above)
Comment (pphaneuf) [inline cpp/fetcher/remote_peer.cc]: Never mind, saw the other one at line 44.

File: cpp/fetcher/remote_peer.cc [hunk 5]
```diff
@@ -3,9 +3,12 @@
 #include <chrono>
 #include <gflags/gflags.h>
 #include <glog/logging.h>
+#include <monitoring/monitoring.h>
```
Comment (pphaneuf) [inline cpp/fetcher/remote_peer.cc]: With `<>`, not `""`?

Comment (AlCutter) [inline cpp/fetcher/remote_peer.cc]: Tired braining I think, fixed.

**Comment by pphaneuf** [comment]
Done with this review.

**Comment by pphaneuf** [comment]
LGTM

File: cpp/fetcher/remote_peer.cc [hunk 6]
```diff
@@ -67,28 +80,62 @@ void RemotePeer::Impl::DoneGetSTH(
...
+    case LogVerifier::INVALID_MERKLE_PATH:
+    case LogVerifier::INVALID_FORMAT:
+      sth_provisionally_valid = false;
+      LOG(FATAL) << "Unexpected verify result for VerifySignedTreeHead(): "
+                 << result;
+  }
+
+  const string local_root_at_snapshot(
+      log_lookup_->RootAtSnapshot(new_sth->tree_size()));
+  if (new_sth->sha256_root_hash() != local_root_at_snapshot) {
```
Comment (pphaneuf) [inline cpp/fetcher/remote_peer.cc]: I've already LGTM'd this, but I came back to check that #761 indeed had everything I wanted...

This implies that we already have the entries for the STH that we received? Isn't the interesting case the one where you have N entries, get a new STH with N+X entries, triggering the fetching of those next X entries?

**Comment by AlCutter** [comment]
Yes, this is actually broken as is, that'll teach me to try to shuffle things around to keep it all nice and "tidy."

Followup PR doing it in the original place heading your way shortly.

**Pull request merged**